### PR TITLE
50 implement switch protocol communicator

### PR DESCRIPTION
**Pull request by LucasTesnier**
Finish the development of the communicator.

Implement and rebase the data transfer protocol :
- Unify the header identification and production process
- Add a new data inside the header (Data type)
- Fix some misunderstanding with the header memory manipulation
- Add malloc verification

Implement data type management :
- Add basic data type (10 -> Connection, 30 -> ECS) (Documentation will be added inside the RFC in future)
- Add intern data type (20 -> Kick/Rebase operation, 21 -> Answer of the 20) (Documentation will be added inside the RFC in future)

Implement functional testing :
- Add a simulation of three Communicator interacting with each other (Some other scripting test will be added in the future)

Fix an oversight inside the communicator utility function:
- Add a bridge between Communicator and sender module to sends some data.

The Communicator is now able to tell another Communicator to change his interlocutor.

### CODE REVIEW

**Review by 31Nathan — APPROVED** [review]
Seems good to me, although I have some difficulties understanding the network

**Comment by osvegn** [comment]
@LucasTesnier
Is there an issue (or some issues) with functionality or documentation you want to implement in the future ? Otherwise, please do it.

File: libs/Communicator/Receiver.hpp [hunk 1]
```diff
@@ -94,6 +97,30 @@ namespace communicator_lib
...
+        /// @param dataContent The transfered data (Client + Data)
+        void dataTraitmentType20(Message dataContent);
+
+        /// @brief Trait the data for the dataType 21
+        /// @param dataContent The transfered data (Client + Data)
+        void dataTraitmentType21(Message dataContent);
+
+        /// @brief Trait the data for the dataType 30
+        /// @param dataContent The transfered data (Client + Data)
+        void dataTraitmentType30(Message dataContent);
```
Comment (osvegn) [inline libs/Communicator/Receiver.hpp:119]: > trait the data

It's not precise enough. It seems these function "add message" somewhere, it's more precise than "trait data".

**Review by osvegn — CHANGES_REQUESTED** [review]
Remove the Client class destructor in the header file and set it to default.
Some documentation to improve.

File: libs/Communicator/Communicator.cpp [hunk 2]
```diff
@@ -91,12 +92,49 @@ void Communicator::kickAClient(Client client, Client newEndpoint)
         return;
     } catch (NetworkError &error) {
     }
-    if (newEndpoint == Client())
-        _senderModule.sendDataToAClient(client, (void *)"kick", 5); /// TO REFACTO WHEN UDP PROTOCOL IS IMPLEMENTED
-    else {
-        unsigned short temp = newEndpoint.getPort();
-        _senderModule.sendDataToAClient(client, &temp, 2); /// TO REFACTO WHEN UDP PROTOCOL IS IMPLEMENTED
+    if (newEndpoint == Client()) {
+        _senderModule.sendDataToAClient(client, (void *)"kick", 5, 30); /// TO REFACTO WHEN UDP PROTOCOL IS IMPLEMENTED
```
[the remark this answers is not in the document]
Comment (LucasTesnier) [reply libs/Communicator/Communicator.cpp:96]: https://github.com/R-Type-Epitech-Nantes/R-Type/issues/60

**Review by 31Nathan — APPROVED** [review]

**Review by osvegn — APPROVED** [review]
All right

**Pull request merged**